README.md

```diff
@@ -72,96 +73,105 @@ Adverbs are "slowly" and "noisily"
 
-  (Hint: my answer has seven states including the start and end states)
+You can hand draw it and snap a picture, or use a drawing tool. Either way, upload the image, and put the file name in the answer below.
 
+(Hint: my answer has seven states including the start and end states)
 
-## A2.3
+A2.3
 
-«replace this with your answer»
+Photo is attached
 
-
-## Q2.4 (6 points)
+Q2.4 (6 points)
 
 Convert this diagram into a table of the form:
 
-Current state | Next word | Next state
---------------|-----------|-----------
-    S0        |    the    |     S1
-    S1        |   . . .   |   . . .
-
-(hint: my version has 13 entries. Yours _might_ be different)
+<<<<<<< HEAD
+Current state Next word Next state
+=======
+Current state	Next word	Next state
+>>>>>>> 07f1444fe496e18ef7278164002acd9890749777
+S0        |    the    |     S1
+S1        |   . . .   |   . . .
+(hint: my version has 13 entries. Yours might be different)
 
-## A2.4
+A2.4
 
-«replace this with your answer»
+<<<<<<< HEAD
+Current    Next (word)  Next (state)
+=======
+Current    Next (word)	Next (state)
+>>>>>>> 07f1444fe496e18ef7278164002acd9890749777
+S0        |    the    |   S1
+S1        | Adjective |   S2
+S1        |    Noun   |   S3
+S2        | Adjective |   S2
+S2        |    Noun   |   S3
+S3        |   Verb    |   S4
+S4        |  Adverb   |   S5
+S4        |    EOI    |   END
+S5        |    EOI    |   END
 
 
-## Q2.5 (12 points)
+Q2.5 (12 points)
 
-Translate this table into a programming language of your choice. Then write a
-function that takes a list of words (ending "EOI") and runs the list through the
-state machine. If the state machine cannot find a transition for a word when in
-a given state, return false. If the state machine runs out of words and the
-state is not "END" then return false. Otherwise return true.
+Translate this table into a programming language of your choice. Then write a function that takes a list of words (ending "EOI") and runs the list through the state machine. If the state machine cannot find a transition for a word when in a given state, return false. If the state machine runs out of words and the state is not "END" then return false. Otherwise return true.
 
-Then write some unit tests that exercise your function, making sure that it
-recognizes valid sentences and rejects invalid ones.
+Then write some unit tests that exercise your function, making sure that it recognizes valid sentences and rejects invalid ones.
 
-The answer should appear in one or more source files in the same directory as
-this file. If I need to do anything more that type a single command to run your
-code, include a script or makefile that will do the job.
+The answer should appear in one or more source files in the same directory as this file. If I need to do anything more that type a single command to run your code, include a script or makefile that will do the job.
 
-## A2.5
+A2.5
 
-«replace this with your answer»
+*See attatched program
```

Comment:
So, a couple of problems. Although you have three pieces of code that exercise your grammar, they don't qualify as tests. A test knows what the expected result should be, and then complains if the actual result is different. There should be no need for human involvement, except perhaps receiving a notification that something is wrong.

# -2

Technically, you code doesn't parse my grammar, because grammar had smelly and lazy dogs and cat, and you've just implemented it in terms of the nonterminals. If we were writing a general purpose language parser, then this kind of classification-based parse might be useful. I think I'm going to let you get away with it. :)

However, I can't overlook the fact that you didn't actually implement a state machine. What you have here is a simple manual parser.  Here's the kind of state machine I was hoping for:

~~~ cpp
typedef int state;
typedef std::map<std::string, state> transitions;
typedef std::map<state, transitions> state_table;

typedef vector<string> sentence;


const int END = 99;

state_table states = {
// current    transitions
// state      {word -> new state}
// ---------+-------------------------------------------------
  {     0,    {{ "the",    1 }}},
	{     1,    {{ "lazy",   2 }, { "smelly",  2 },   { "dog", 3 }, {"cat", 3 }}},
	{     2,    {{ "lazy",   2 }, { "smelly",  2 },   { "dog", 3 }, {"cat", 3 }}},
	{     3,    {{ "ate",    4 }, { "ran",     4 }}},
	{     4,    {{ "slowly", 5 }, { "noisily", 5 },   { "EOI", END }}},
	{     5,    {{ "EOI",    END }}}
};


int getNextState(int currentState, string nextWord) {
	transitions t = states[currentState];
	if (t.count(nextWord))
		return t[nextWord];
	else
		return -1;
}

bool validSentence(sentence words) {
    int state = 0;
    for (string word: words) {
        state = getNextState(state, word);
        if (state < 0)
            return false;
    }
    return state == 5;
}
~~~

# -6